pkg/apis/flowcontrol/validation/validation.go

```diff
@@ -0,0 +1,296 @@
+		allErrs = append(allErrs, ValidateFlowSchemaResourcePolicyRule(&resourceRule, fldPath.Child("resourceRules").Index(i))...)
+	}
+	for i, nonResourceRule := range rule.NonResourceRules {
+		allErrs = append(allErrs, ValidateFlowSchemaNonResourcePolicyRule(&nonResourceRule, fldPath.Child("nonResourceRules").Index(i))...)
+	}
+	return allErrs
+}
+
+// ValidateFlowSchemaSubject validates flow-schema's subject object.
+func ValidateFlowSchemaSubject(subject *flowcontrol.Subject, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	switch subject.Kind {
+	case flowcontrol.SubjectKindServiceAccount:
+		allErrs = append(allErrs, ValidateServiceAccountSubject(subject.ServiceAccount, fldPath.Child("serviceAccount"))...)
+	case flowcontrol.SubjectKindUser:
+		allErrs = append(allErrs, ValidateUserSubject(subject.User, fldPath.Child("user"))...)
+	case flowcontrol.SubjectKindGroup:
+		allErrs = append(allErrs, ValidateGroupSubject(subject.Group, fldPath.Child("group"))...)
+	default:
+		allErrs = append(allErrs, field.NotSupported(fldPath.Child("kind"), subject.Kind, supportedSubjectKinds.List()))
+	}
+	return allErrs
+}
+
+// ValidateServiceAccountSubject validates subject of "ServiceAccount" kind
+func ValidateServiceAccountSubject(subject *flowcontrol.ServiceAccountSubject, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if len(subject.Name) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("name"), ""))
+	}
+	if len(subject.Name) > 0 && subject.Name != flowcontrol.NameAll {
+		for _, msg := range validation.ValidateServiceAccountName(subject.Name, false) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("name"), subject.Name, msg))
+		}
+	}
+	if len(subject.Namespace) > 0 {
+		for _, msg := range apimachineryvalidation.ValidateNamespaceName(subject.Namespace, false) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("namespace"), subject.Namespace, msg))
+		}
+	} else {
+		allErrs = append(allErrs, field.Required(fldPath.Child("namespace"), "must specify namespace for service account"))
+	}
+	return allErrs
+}
+
+// ValidateUserSubject validates subject of "User" kind
+func ValidateUserSubject(subject *flowcontrol.UserSubject, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if len(subject.Name) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("name"), ""))
+	}
+	return allErrs
+}
+
+// ValidateGroupSubject validates subject of "Group" kind
+func ValidateGroupSubject(subject *flowcontrol.GroupSubject, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if len(subject.Name) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("name"), ""))
+	}
+	return allErrs
+}
+
+// ValidateFlowSchemaNonResourcePolicyRule validates non-resource policy-rule in the flow-schema.
+func ValidateFlowSchemaNonResourcePolicyRule(rule *flowcontrol.NonResourcePolicyRule, fldPath *field.Path) field.ErrorList {
```

Comment:
This is not checking for URL prefixes that the comment says are illegal --- but should.

Reply:
i dont see such validation in rbac, are we starting that from this pull? i think it will involve more issues.

Reply:
Are there difficult issues in enforcing the comments?

We can leave some non-interface-changing work for later PRs, but I think we will not be done until the comments match the validation logic and the handling by the implementation.